resources/type-system/inference.md

```diff
@@ -1134,25 +1134,58 @@ succintly, the syntax of Dart is extended to allow the following forms:
   any loss of functionality, provided they are not trying to construct a proof
   of soundness._
 
-In addition, the following forms are added to allow constructor invocations,
-dynamic method invocations, function object invocations, instance method
-invocations, and static/toplevel method invocations to be distinguished:
+In addition, forms are added to allow constructor invocations, dynamic method
+invocations, function object invocations, instance method invocations, and
+static/toplevel method invocations to be distinguished. In these forms, `n_i:
+m_i` (where `i` is an integer) is used as a convenient meta-syntax to refer to
+an invocation argument `m_i` (an elaborated expression), possibly preceded by a
+name selector `n_i:` (where `n_i` is a string). In this document, we use the
+string `∅` to represent a name selector which is absent (meaning the
+corresponding `m_i` is a positional argument rather than a named argument).
+
+The new invocation forms are as follows:
 
 - `@DYNAMIC_INVOKE(m_0.id<T_1, T_2, ...>(n_1: m_1, n_2: m_2, ...))`
 
+  _This covers invocations of user-definable binary and unary operators, method
+  invocations, and implicit uses of the `call` method, where the target has
+  static type `dynamic` or `Function`. For example, if `d` has static type
+  `dynamic`, the following expressions will be elaborated using
+  `@DYNAMIC_INVOKE`: `d.f()`, `-d`, `d + 0`, and `d()`._
+
 - `@FUNCTION_INVOKE(m_0.call<T_1, T_2, ...>(n_1: m_1, n_2: m_2, ...))`
 
+  _This covers invocations of expressions whose type is a function type (but not
```

Comment:
I don't really understand the division between FUNCTION_INVOKE and STATIC_INVOKE.  At a minimum, I think you need to say something here like "expressions which do not resolve to top level or static methods" because otherwise it is ambiguous what to do with `f(3)` where `f` is a top level method (because `f` is an expression whose type is a function type, and it is also a top level method).  I'm also confused about something like `x.call()`.  I guess maybe the screwy Dart grammar doesn't treat the `x.call` here as an expression and therefore we can say that it is unambiguously covered by INSTANCE_INVOKE instead of being ambiguous?  Is that right?  So `(x.call)()` is a FUNCTION_INVOKE, but `x.call()` is an INSTANCE_INVOKE?